spi-proposal.md

```diff
@@ -0,0 +1,354 @@
+
+### SPI Config Device
+
+Send once for each SPI device, specifying a new deviceId for each device
+(beginning from 0).
+
+#### bitOrder
+```
+LSBFIRST = 0
+MSBFIRST = 1
+```
+
+#### dataModes
+
+| mode    | clock polarity (CPOL) | clock phase (CPHA) |
+| --------|-----------------------|--------------------|
+| 0       | 0                     | 0                  |
+| 1       | 0                     | 1                  |
+| 2       | 1                     | 0                  |
+| 3       | 1                     | 1                  |
+
+
+#### optionalPins
+
+The `optionalPins` byte allows HW pins such as chip select, data ready, and
+other common (or perhaps even uncommon) actions realated to SPI HW that are
+not directly handled by SPI libraries. Optional pins are identified as bits of
+the `optionalPins` byte.
+
+```
+bit 0 = chip select
+bit 1 = data ready
+... additional bits TBD as needed
+```
+If one or more bits are set, then the pin number and an options byte are
+defined for each pin (so 2 bytes per pin) in order beginning with bit 0.
+
+The optional pins are described below:
+
+```
+Chip select (CS):
+0:  csPin                (0-127)
+1:  csOptions:
+      csActive (bit 0)   (0 = LOW -> HIGH, 1 = HIGH -> LOW, default = 0)
+      csToggle (bit 1)   (0 = toggle after last byte,
+                          1 = toggle between bytes, default = 0)
+
+Data ready:
+0:  dataReadyPin         (0-127)
+1:  dataReadyOptions:
+      dataReadyActive (bit 0) (0 = LOW, 1 = HIGH, default = 1)
+```
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_CONFIG_DEVICE    (0x00)
+3:  deviceId | channel   (bits 2-6: deviceId, bits 0-1: channel (0-3))
+4:  bitOrder | dataMode  (bit 2: bitOrder, bits 0-1: dataMode (0-3))
+                          see mode column in table above
+5:  clockSpeed           (bits 0 - 6)
+6:  clockSpeed           (bits 7 - 14)
+7:  clockSpeed           (bits 15 - 21)
+8:  clockSpeed           (bits 22 - 28)
+9:  clockSpeed           (bits 29 - 32)
+10: optionalPins         (see description above)
+... defined optional pins (2 bytes each)
+N:  END_SYSEX (N = 10 + (num optional pins * 2))
+```
+
+### SPI Transfer
+
+**Read a bybe for each byte written.**
+
+Send an array of data to transfer to the SPI slave device, expecting a byte
+returned for each byte written. Respond with a `SPI_REPLY` message.
+
+Use the **sequence** byte to specify if this is a one time transfer or if it is part
+of a sequence of transfers, or an ongoing stream of transfers. Define as
+follows:
+
+```
+0 = single transfer (no sequence)
+1 = start sequence (1st part of sequence)
+2 = 2nd part of sequence
+3 = 3rd part of sequence, etc...
+... up to 100, after 100, wrap back to 1
+127 = end sequence
+```
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_TRANSFER_REQUEST (0x01)
+3:  deviceId | channel   (bits 2-6: deviceId, bits 0-1: channel)
+4:  sequence             (see note on sequence above)
+5:  numBytes to read and write
```

Comment:
Here's an example (copied and modified from [here](https://www.arduino.cc/en/Tutorial/BarometricPressureSensor)) of what I mean by multi-byte:

```c++
digitalWrite(chipSelectPin, LOW);
SPI.transfer(thisRegister);
result = SPI.transfer(0x00);
bytesToRead--;
if (bytesToRead > 0) {
    result = result << 8;
    inByte = SPI.transfer(0x00);
    result = result | inByte;
    bytesToRead--;
}
digitalWrite(chipSelectPin, HIGH);
```

On the 2nd line, a value is written via SPI.transfer, then 2 bytes are read (where a value of 0x00 is written since you must always write a value to return a value). However the tricky part with a Firmata implementation is how to handle situations like this in a general manner. The slowest way is to send a separate message for each step (where max rate of transfer is = baud if using a Serial transport):

```js
// from client side
FirmataClient.spi.beginTransaction(clockSpeed, dataOrder, dataMode)
FirmataClient.sendDigitalMessage(csPin, LOW)
// W = write only, R = read only, RW = read & write
FirmataClient.spi.transfer(W, registerValue)
FirmataClient.spi.transfer(RW, 0x00, callback)
... wait for return value
FirmataClient.spi.transfer(RW, 0x00, callback)
... wait for return value
FirmataClient.sendDigitalMessage(csPin, HIGH)
FirmataClient.spi.endTransaction()
```

That would require at least 7 separate messages including wait time in between those that expect a reply. However if I can figure out a way to generically handle writing and reading streams, it would be far more efficient:

```js
// from client side
// only need to send config once per device
FirmataClient.spi.config(deviceId, clockSpeed, dataOrder, dataMode, csPin, csOptions)
// this single transfer would perform the same 7 steps as the previous example, but
// all on the firmware side rather than on the client side and at MCU clock speed
// rather than at baud rate
FirmataClients.spi.transfer(deviceId, sequence, numBytes, dataArray)
... wait for reply (which should have all of the data in a single array
```

While it may not work for every case, this approach could accomplish the same as the first example in a single message over the wire vs 7 separate messages. The only wait necessary is for the reply, which should contain all of the requested data.